### DISCUSSION TITLE
Argon2 invalid salt size when specifying a different cost hash

### DISCUSSION THREAD

**Discussion by estebanz01** [jeremyevans/rodauth#349] [General]
Hola de nuevo!

so right now I'm in a bit of a pickle trying to login with a password that is being hashed as argon2 and custom cost hash. I specified it because I need to match the same values used in django for my app written in ruby. I'm getting the following error:

```ruby
 $ curl -X POST http://localhost:3000/api/v1/login -H 'Content-Type: application/json' -d '{"email": "adios@example.com", "password": "holamundo123holamundo"}'
Argon2::ArgonHashFail: Invalid salt size
	/home/esteban/.local/gel/3.2.0/gems/argon2-2.2.0/lib/argon2/ffi_engine.rb:85:in `hash_argon2id_encode'
	/home/esteban/.local/gel/3.2.0/gems/argon2-2.2.0/lib/argon2.rb:34:in `create'
	/home/esteban/.local/gel/ruby/gems/rodauth-2.30.0/lib/rodauth/features/argon2.rb:60:in `password_hash_using_salt'
	/home/esteban/.local/gel/ruby/gems/rodauth-2.30.0/lib/rodauth/features/base.rb:519:in `database_function_password_match?'
	/home/esteban/.local/gel/ruby/gems/rodauth-2.30.0/lib/rodauth/features/base.rb:434:in `password_match?'
	/home/esteban/.local/gel/ruby/gems/rodauth-2.30.0/lib/rodauth/features/login.rb:66:in `block (4 levels) in <module:Rodauth>'
	/home/esteban/.local/gel/ruby/gems/rodauth-2.30.0/lib/rodauth/features/base.rb:606:in `catch'
	/home/esteban/.local/gel/ruby/gems/rodauth-2.30.0/lib/rodauth/features/base.rb:606:in `catch_error'
	/home/esteban/.local/gel/ruby/gems/rodauth-2.30.0/lib/rodauth/features/login.rb:44:in `block (3 levels) in <module:Rodauth>'
```

Here's the generated hash for a cost hash `t_cost: 2, m_cost: 16, p_cost: 8`: `$argon2id$v=19$m=65536,t=2,p=8$5cUrI7nWSzKldLW7JxYxfw$itpqsB10lQVTLm8qUDOJQI8zhw8JXVbEBIVRXBWZxgY`

and for a cost hash `t_cost: 2, m_cost: 20, p_cost: 8`: `$argon2id$v=19$m=1048576,t=2,p=8$ldSO8VFvjdhSrIpaRclRxw$/f47QmzFyYERDSSNwNpwNKOskgFRHghTb1M/DWDEgTk`

if I go to the DB and call the salt function, I get the following:
```SQL
db=# SELECT rodauth_get_salt(3);
       rodauth_get_salt        
-------------------------------
 $argon2id$v=19$m=1048576,t=2,
(1 row)

db=# SELECT rodauth_get_salt(4);
       rodauth_get_salt        
-------------------------------
 $argon2id$v=19$m=65536,t=2,p=
(1 row)
```

and I can confirm that the user that rodauth is using to connect to the DB has access and can query the `rodauth_get_salt` method.

I'm really lost on how to debug this, so any guidance will be really appreciated !

Thanks!

```ruby
# Argon2 rodauth configuration
argon2_secret ENV.fetch('SECRET_VALUE') { SecureRandom.hex(256) }
require_bcrypt? false
# Match cost configuration with default Django
password_hash_cost({ t_cost: 2, m_cost: 16, p_cost: 8 })
```

**Comment by jeremyevans** [comment]
Did you pass the `:argon2` option when calling `Rodauth.create_database_authentication_functions`?  My guess would be no, considering the returned values are 29 characters, which is what Rodauth returns for bcrypt hashes.  See the argon2 feature documentation for how to fix this: https://rodauth.jeremyevans.net/rdoc/files/doc/argon2_rdoc.html .  Can you try that and report back whether it works for you?

**Reply by estebanz01** [reply]
Ah ok, since I'm using postgresql, it's not clear if I need to recreate or not the functions. I'll give it a try and report back.

**Reply by estebanz01** [reply]
OK, I nuked the DB, added the `argon2: true` parameter to the function creations part and I'm getting a weird behaviour. Having the following:

```ruby
class Rodauth < Roda
  plugin :middleware
  plugin :rodauth, json: :only do
    enable :login, :logout, :create_account, :active_sessions, :json, :jwt, :jwt_refresh, :argon2

    hmac_secret ENV.fetch('SESSION_SECRET')
    jwt_secret ENV.fetch('SESSION_SECRET')

    argon2_secret 'español'
    require_bcrypt? false
end
```

when I create an user with a curl request, it works and I can logout/login again as that user:

```bash
> curl -X POST http://localhost:3000/api/v1/create-account -H 'Content-Type: application/json' -d '{"email": "adios@example.com", "password": "holamundo123holamundo", "password-confirm": "holamundo123holamundo", "login-confirm": "adios@example.com"}'
{"access_token":"<access-token>","refresh_token":"<refresh-token>","success":"Your account has been created"}

# saved argon2 hash: $argon2id$v=19$m=65536,t=2,p=1$78corUjukKeQ+xEp2Jp4GA$bnhvYXWI8QZqyQQiknNG6qrbH7fMfTd76uoSvA9739Q

# logout then, login

> curl -X POST http://localhost:3000/api/v1/login -H 'Content-Type: application/json' -d '{"email": "hola@example.com", "password": "holamundo123holamundo"}'
{"access_token":"<access-token>","refresh_token":"<refresh-token>","success":"You have been logged in"}
```

now, if I stop the server and add the following configuration for argon2:

`password_hash_cost({ t_cost: 2, m_cost: 20, p_cost: 8 })`

I can create a new user, but once I logout, I cannot login properly:

```bash
> curl -X POST http://localhost:3000/api/v1/create-account -H 'Content-Type: application/json' -d '{"email": "adios@example.com", "password": "holamundo123holamundo", "password-confirm": "holamundo123holamundo", "login-confirm": "adios@example.com"}'
{"access_token":"<access-token>","refresh_token":"<refresh-token>","success":"Your account has been created"}

# saved argon2 hash: $argon2id$v=19$m=1048576,t=2,p=8$kId8dDrxQyMxX7jDKxqQ9g$Y9RsoFENMCRhx4ng1pGBYV5OBUPEUmOr6ZVJTyWMJEI

# Logout without problems, error when trying to login

> curl -X POST http://localhost:3000/api/v1/login -H 'Content-Type: application/json' -d '{"email": "adios@example.com", "password": "holamundo123holamundo"}'
{"field-error":["password","invalid password"],"error":"There was an error logging in"}
```

and if I try to login with the user that was created with the default hash cost, it works.

Now, if I wipe the entire DB again and create both users _with the changed password hash cost_, I cannot login with both users either.

am I missing something else in the configuration or am I doing something wrong in it?

**Reply by estebanz01** [reply]
btw, here's my test with the argon2 gem directly on an IRB console to verify the input with an argon2 secret:

```ruby
3.2.2 :001 > require 'argon2'
 => true 
3.2.2 :002 > Argon2::Password.new(t_cost: 2, m_cost: 20, p_cost: 8, secret: 'español').create('holamundo123holamundo')
 => "$argon2id$v=19$m=1048576,t=2,p=8$rXantnC0I+M4H2FxLQ8WNQ$FwCrmpdBxC3RPxMIX4/Za0Cx4eMiLUTTJZUHiawGXGw" 
3.2.2 :003 > Argon2::Password.verify_password("holamundo123holamundo", "$argon2id$v=19$m=1048576,t=2,p=8$rXantnC0I+M4H2FxLQ8WNQ$FwCrmpdBxC3RPxMIX4
/Za0Cx4eMiLUTTJZUHiawGXGw", 'español')
 => true 
3.2.2 :004 > Argon2::Password.new(t_cost: 2, m_cost: 16, p_cost: 1, secret: 'español').create('holamundo123holamundo')
 => "$argon2id$v=19$m=65536,t=2,p=1$sBvDsGEOdWajvAmtuKbUCw$zmi6sDXtmJZ1sNLWjk+YPFY07ar3+Pvif6BqPSzYYbc" 
3.2.2 :005 > Argon2::Password.verify_password("holamundo123holamundo", "$argon2id$v=19$m=65536,t=2,p=1$sBvDsGEOdWajvAmtuKbUCw$zmi6sDXtmJZ1sNLWjk+Y
PFY07ar3+Pvif6BqPSzYYbc", "español")
 => true
```

**Reply by jeremyevans** [reply]
I tested and had no problems:

```ruby
# t.ru
require 'roda'
require 'sequel'
DB = Sequel.postgres(:user=>'rodauth_test')

class RodauthTest < Roda
  plugin :common_logger
  plugin :rodauth, json: :only do
    secret = '1'*64
    enable :login, :logout, :create_account, :active_sessions, :json, :jwt, :jwt_refresh, :argon2

    hmac_secret secret
    jwt_secret secret
    require_login_confirmation? false
    require_password_confirmation? false
    login_param 'email'
    # password_hash_cost({ t_cost: 2, m_cost: 20, p_cost: 8 }) # uncomment after creating first user

    argon2_secret 'español'
    require_bcrypt? false
  end

  route do |r|
    r.on 'api/v1' do
      r.rodauth
    end
  end
end

run RodauthTest.app.freeze
```

With the above configuration, I could create and login to the account with the default cost.  After changing the cost, I could login to the already created account that used the default cost, and also create and login to a new account (which took much longer due to the increased cost).

In the future, please provide a tested self contained example like the one above, and tested example code.  Your example code provided didn't work with the configuration you provided, and it took a bit a guesswork to determine what configuration would work.  Even with the configuration fixed, your `curl` examples didn't work because the first set create a user for `adios@example.com` and then try to login with `hola@example.com`, and the second set try to create the same account used in the first set (which should fail unless you use a different account or wipe the database in between)

I'm not sure why it is failing for you, but it appears to be a local issue.

**Reply by estebanz01** [reply]
OK, I'll keep debugging. Thanks for the contained example. Re curl commands, probably a copy/paste issue as I posted after a long day of work.

**Reply by estebanz01** [reply]
I found my issue. The `p_cost` is not being passed when checking if the password hash matches or not: https://github.com/jeremyevans/rodauth/blob/master/lib/rodauth/features/argon2.rb#L67, I manually changed the regex and the result to be:

```ruby
/\A\$argon2id\$v=\d+\$m=(\d+),t=(\d+),p=(\d+)/ =~ hash
{ t_cost: $2.to_i, m_cost: Math.log2($1.to_i).to_i, p_cost: $3.to_i }
```

and the login works as expected. Not sure if this is a bug or the intended behaviour is to ignore `p_cost`, in any case, I have decided to go with defaults, which are good enough.

hash generated before my change: `$argon2id$v=19$m=1048576,t=2,p=1$Cv3Kae1xTcSdrv1xyUajfg$N1KmexbCHR0oHrSOrZoXrU35nnGK6LeknPxnrafGHSU`

and the hash generated after my change:
`$argon2id$v=19$m=1048576,t=2,p=8$Cv3Kae1xTcSdrv1xyUajfg$4AZWAgwCTysWJ9t99KF1KhDHdhFRusxoZl7WrygtwGA`

notice the `p_cost` as `8`.

**Reply by jeremyevans** [reply]
Rodauth doesn't set the default `p_cost`, but I'm open to recognizing it and setting it on extraction (in `extract_password_hash_cost`).  Do you want to send in a pull request for that?

**Reply by estebanz01** [reply]
sure thing
